# The Smith-Waterman Algorithm
The Smith-Waterman algorithm is a dynamic programming algorithm to find the optimal alignment of two sequences.

This repository provides a basic implementation of the Smith-Waterman algorithm that has been tailored to the issue of finding the optimal alinment of two DNA protein sequences. Sequence aligment is scored via the BLOSUM62 scoring matrix. The BLOSUM62 scoring matrix is the default scoring matrix for [BLASTP](https://blast.ncbi.nlm.nih.gov/Blast.cgi) and was developed by analyzing the frequencies of amino acid substitutions in clusters of related proteins.

*note: this algorithm was implemented for an assigment in a computation biology class I took in early 2021.*

### A Short Primer on DNA Sequence Alignment
<todo: writeup>

### Why Dynamic Programming?
<todo: writeup>

### Precursor: The Needleman-Wunsch Algorithm
<todo: writeup>

### How Smith-Waterman Works
<todo: writeup>

